Dont mutate tensor stride in place in cudnn conv

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126786

Fix for https://github.com/pytorch/pytorch/issues/126241. 

Within the cudnn convolution, we were in-place updating the strides of the tensor to disambiguate for size-1 dims and contiguous and channels last tensors. Instead of mutating the tensors stride, just use a temporary. Inside cudnn it is then copied: https://github.com/NVIDIA/cudnn-frontend/blob/d7ccb5b3c47b4de709604cce463ad66b775b7812/include/cudnn_frontend_Tensor.h#L201-L203.

eellison committed:
Dont mutate tensor stride in place in cudnn conv

[ghstack-poisoned]

eqy reviewed (commented):
Thanks!

Does this need to also be fixed for legacy API use-cases that are creating `TensorDescriptor`s by passing `Tensor`s directly and calling `.stride()` on them? e.g., https://github.com/pytorch/pytorch/blob/d8f5627a88167cbbbd64c2b98b67a4dee64ce211/aten/src/ATen/cudnn/Descriptors.cpp#L38

eellison commented:
What are the callsites of those apis ?

eellison commented:
We don't compile RNNs in torch.compile - would prefer separate issue/fix for the other legacy callsites.

eqy commented:
@eellison Yes I think overloads of `set()` bottom out on the variant here https://github.com/pytorch/pytorch/blob/b40fb2de5934afea63231eb6d18cc999e228100f/aten/src/ATen/cudnn/Descriptors.h#L169 that calls `fixSizeOneDimStride`. The callsites are IIRC RNN, Conv v7, and maybe other legacy API places like batchnorm?

eellison committed:
Update on "Dont mutate tensor stride in place in cudnn conv"


Fix for https://github.com/pytorch/pytorch/issues/126241. 

Within the cudnn convolution, we were in-place updating the strides of the tensor to disambiguate for size-1 dims and contiguous and channels last tensors. Instead of mutating the tensors stride, just use a temporary. Inside cudnn it is then copied: https://github.com/NVIDIA/cudnn-frontend/blob/d7ccb5b3c47b4de709604cce463ad66b775b7812/include/cudnn_frontend_Tensor.h#L201-L203.



[ghstack-poisoned]

eellison committed:
Update on "Dont mutate tensor stride in place in cudnn conv"


Fix for https://github.com/pytorch/pytorch/issues/126241. 

Within the cudnn convolution, we were in-place updating the strides of the tensor to disambiguate for size-1 dims and contiguous and channels last tensors. Instead of mutating the tensors stride, just use a temporary. Inside cudnn it is then copied: https://github.com/NVIDIA/cudnn-frontend/blob/d7ccb5b3c47b4de709604cce463ad66b775b7812/include/cudnn_frontend_Tensor.h#L201-L203.



[ghstack-poisoned]

ezyang reviewed (approved):
This is fine as bandaid but better would be to refactor call sites

eellison commented:
@pytorchbot merge
